runtime/pkg/rilltime/rilltime_test.go

```diff
@@ -1,176 +1,576 @@
+		{"2024-12-01T00:00:00Z", "2023-03-01T00:00:00Z"}, // for M
+		{"2023-07-01T00:00:00Z"},                         // for Q
+	}
+
+	var grainParis []testGrainPair
+
+	for i, grain := range grains {
+		if i < len(grains)-1 {
+			grainParis = append(grainParis, testGrainPair{grain, grains[i+1]})
+		}
+		if i < len(grains)-2 {
+			grainParis = append(grainParis, testGrainPair{grain, grains[i+2]})
+		}
+	}
+	// Select tests with higher order grains
+	//grainParis = append(grainParis, testGrainPair{"D", "Y"})
+	//grainParis = append(grainParis, testGrainPair{"W", "Y"})
+
+	for _, grainPair := range grainParis {
+		grainIndex := slices.Index(grains, grainPair.grain)
+		higherGrainIndex := slices.Index(grains, grainPair.higherGrain)
+		indexDiff := higherGrainIndex - grainIndex - 1
+		tmGrain := timeutil.TimeGrain(grainIndex + 2)
+
+		snapGrain := grainPair.higherGrain
+		if grainPair.grain == "W" {
+			snapGrain += "W"
+		}
+
+		lastStart := lastStarts[grainIndex][indexDiff]
+		ordinalStart := ordinalStarts[grainIndex][indexDiff]
+		firstEnd := firstEnds[grainIndex][indexDiff]
+
+		higherPeriod := twoPeriodAgoStarts[higherGrainIndex]
+		if grainPair.grain == "W" {
+			higherPeriod = twoPeriodAgoWeekBoundaryStarts[higherGrainIndex]
+		}
+
+		// Last 2 periods of -2<higher grain>
+		testCases = append(testCases, testCase{
+			timeRange: fmt.Sprintf("2%[1]s ending -2%[2]s/%[3]s^", grainPair.grain, grainPair.higherGrain, snapGrain),
+			start:     lastStart,
+			end:       higherPeriod,
+			grain:     tmGrain,
+		})
+		testCases = append(testCases, testCase{
+			timeRange: fmt.Sprintf("-2%[2]s/%[3]s^-2%[1]s to -2%[2]s/%[3]s^", grainPair.grain, grainPair.higherGrain, snapGrain),
+			start:     lastStart,
+			end:       higherPeriod,
+			//grain:     tmGrain, // TODO: certain cases dont correct identify grain
+		})
+		testCases = append(testCases, testCase{
+			timeRange: fmt.Sprintf("-3%[2]s/%[3]s$-2%[1]s to -3%[2]s/%[3]s$", grainPair.grain, grainPair.higherGrain, snapGrain),
+			start:     lastStart,
+			end:       higherPeriod,
+			//grain:     tmGrain, // TODO: certain cases dont correct identify grain
+		})
+		testCases = append(testCases, testCase{
+			timeRange: fmt.Sprintf(">2%[1]s of -3%[2]s!", grainPair.grain, grainPair.higherGrain, snapGrain),
+			start:     lastStart,
+			end:       higherPeriod,
+			grain:     tmGrain,
+		})
+
+		// First 2 periods of -2<higher grain>
+		testCases = append(testCases, testCase{
+			timeRange: fmt.Sprintf("2%[1]s starting -2%[2]s/%[3]s^", grainPair.grain, grainPair.higherGrain, snapGrain),
+			start:     higherPeriod,
+			end:       firstEnd,
+			grain:     tmGrain,
+		})
+		testCases = append(testCases, testCase{
+			timeRange: fmt.Sprintf("-2%[2]s/%[3]s^ to -2%[2]s/%[3]s^+2%[1]s", grainPair.grain, grainPair.higherGrain, snapGrain),
+			start:     higherPeriod,
+			end:       firstEnd,
+			//grain:     tmGrain, // TODO: certain cases dont correct identify grain
```

Comment:
Does this need to be addressed?

Reply:
This returns the grain we show by default. I think this is not that big a deal right now. There will definitely be a follow up once users start using this, we can address this then.